### ISSUE TITLE
Permeability and stencils refactoring idea.

### ISSUE THREAD

**Issue by CusiniM** [GEOSX/GEOSX#1312]
Working on adding pEDFM (#1307) and on adding fractures to the compositional solver (#1311) a few issues came up with the way the permeability and the stencil weights are used. In particular, the current design makes the same assumptions:

- permeability is considered to never change and stencil weights are never updated;
- for the `faceElementStencil` the part of the transmissiblity that depends on the aperture is updated directly in the SinglePhaseFVMKernels which forces us to have a different flux assembly for the `faceElementStencil` (https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVMKernels.cpp#L219-316) and the `cellElementStencil` (https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseFVMKernels.cpp#L125-215).

I think there is a number of use cases in which the transmissiblity can change throughout the simulation both for the fractures and for the matrix cells. Thus, I think it would be good to allow for our stencil weights to be updated based on the primary variables. If we did this we could include the aperture information in the `faceElementStencil` weights and avoid having separate flux kernels for the two stencil types. 

I am proposing to do the following:
- introduce a `constitutiveModel` for the permeability that updates the permeability based on pressure, displacement, aperture (or whatever other variable)
- introduce methods to update the stencil weights
- store the derivatives of the weights w.r.t. to the primary variables on the stencils (or somewhere else but it would be hard to store them on the `subRegions`)
- have only one `fluxKernel` for each flow solver that works for both stencil types.

@klevzoff, @francoishamon, @rrsettgast, @ishovkun , @castelletto1, @joshua-white

**Comment by klevzoff** [comment]
I agree, this area needs redesign. I'm on board with a constitutive model for permeability, although I wish we first figured out how to handle porosity updates. Some thoughts/questions:

1. IMO the stencil should compute and store the derivatives of weights w.r.t. coefficient (e.g. permeability), ***dw/dk***, whereas the permeability model would compute ***dk/dp*** and/or ***dk/du*** (***dk/dA***), and finally the solver would piece them together in a chain rule during assembly (or as a pre-assembly step). Thus a flow-only solver would be able to utilize a (surrogate) permeability model ***k(p)***, while a poromechanical solver could use a model with ***k(u)*** or ***k(A)***. There are also popular models for matrix permeability that model ***k(phi)*** - do these fit anywhere?

2. Question arises as to how complex the dependence should be. For example, when dealing with anisotropic permeability tensors, do we account separately for changes in ***kx***, ***ky*** and ***kz*** (and their derivatives)? If the answer is yes, there may not be a simple way to update the weights, the update would essentially be a full recomputation (with the indices kept) - this may not be a problem, as long as we can do it adaptively and avoid unnecessary recomputation when a static permeability field is used. However, this also means that ***dw/dk*** and ***dk/dp*** are both rank-1 tensors and ***dk/du*** is a rank-2 tensor - this is an increase in memory footprint/traffic that we should at least acknowledge. Ideally it should also be disabled when static perm is used.

3. What would be the driver of permeability and stencil updates? In other words, which solver/component (and when) is responsible for calling `update()` on
   * permeability model
   * `Stencil` or `FluxApproximation` objects

**Comment by CusiniM** [comment]
1. I agree with @klevzoff about how to store the derivatives. I think we should also include Carman-Kozeny type of equations to update the permeability as well.

2. IMO we should consider the possibility of having "complex" relationships in which changes are different for each component of the permeability. Maybe we will have to allow for various options (different constitutive models) and size the objects accordingly as done for the mechanics moduli. 

3. My guess is that we should treat this like a fluid or a solid property update. I mean, every time we update the state variables (so at every iterations) all constitutive models should update the respective parameters. I tend to like the way it's done in the compositional solver at the moment.

**Comment by francoishamon** [comment]
Sounds like a good plan to me ! If we decide to unify the flow kernels, how are we going to handle the "old" case in which the transmissibilities are constant throughout the simulation? 
- Are we going to always pass all the derivatives of transmissibility (even if they are zero) to the `FluxKernel::launch` function and add these zero values into the Jacobian?
- Or are we going to have a switch before the launch to decide whether we want to pass the derivatives of transmissibility to the `FluxKernel::launch` function or not?
- Or should we duplicate the kernels to handle the different cases: constant transmissibilities, transmissibilities with derivatives wrt to pressure, wrt to displacement (in which case the size of the local jacobian changes)?

I believe this brings us back to the discussion we had for the thermal work (but I don't remember what we decided ;) ).

**Comment by ishovkun** [comment]
One thing to keep in mind is that permeability updates should be chained: e.g. an update from matrix deformation followed by a model from pore clogging due to fines transport, followed by a reactive dissolution updated, etc. 

Please keep that in mind and try to incorporate into the design since these features will be a lot harder to incorporate when the codebase is larger.

**Comment by CusiniM** [comment]
Ok, I am going to start adding this design in  (#1311) because it just makes sense to do so when adding fractures to the compositional solver since it will simplify its implementation quite a bit. Also, fractures really need a non-constant permeability.

I only have one question about who should own the permeability. Atm, it is registered on the `subregions` but I think that, having a constitutive model, it would make sense to make the permeability model own it, like we do for most fluid properties in the `MultiphaseMulticomponentFluid`.  My only worry is whether this is going to create any issue when importing an external mesh with some data on it. Will we be able to populate the constitutive model permeability with data from from the imported mesh? If not, would this require major changes?

**Comment by klevzoff** [comment]
> My only worry is whether this is going to create any issue when importing an external mesh with some data on it. Will we be able to populate the constitutive model permeability with data from from the imported mesh? If not, would this require major changes?

Every constitutive data array contained within a model also has a duplicate wrapper under subregions, with the name that looks like `<model_name>_<field_name>`. So if we take an import example like this
https://github.com/GEOSX/GEOSX/blob/b204d2c528d59533929bca7f7a06b4946b8b2da2/src/coreComponents/physicsSolvers/fluidFlow/benchmarks/Egg/dead_oil_egg.xml#L151-L155
and you add a permeability model named `rock`, you can then specify `fieldNamesInGEOSX="{ rock_permeability }` and it ***should*** find the wrapper on the subregion.

However, it will likely not be able to import the data still, because of the extra gauss point dimension that PAMELA importer may not be able to understand (it will be a 3D-array, while the source is 2D). There will be some work involved in teaching `PAMELAMeshGenerator` to handle these cases.

Last but not least: at present, property import happens in the same call as mesh generation (same method in `PAMELAMeshGenerator`). We will need to decouple them, to make sure data import occurs *after* constitutive models have been allocated. I ***hope*** it is as simple as adding a new virtual function in `MeshGeneratorBase` that is empty for internal generator and does the import for PAMELA, and making sure it's called at the right time in `ProblemManager::problemSetup`.

**Comment by klevzoff** [comment]
Would be great to have property import changes handled as a separate PR. Can it be tested with `LinearElasticIsotropic` and its `bulkModulus` field?
